BedrockServer.cpp

```diff
@@ -1004,21 +1015,56 @@ BedrockServer::BedrockServer(const SData& args)
     // Enable the requested plugins, and update our version string if required.
     list<string> pluginNameList = SParseList(args["-plugins"]);
     vector<string> versions = {_version};
+
+    // A map of plugins and what secure data entries they need.
+    map<BedrockPlugin*, list<string>> pluginSecureDataMap;
     for (string& pluginName : pluginNameList) {
         BedrockPlugin* plugin = registeredPluginMap[SToLower(pluginName)];
         if (!plugin) {
             SERROR("Cannot find plugin '" << pluginName << "', aborting.");
         }
-        plugin->initialize(args, *this);
-        plugins.push_back(plugin);
-
-        // If the plugin has version info, add it to the list.
-        auto info = plugin->getInfo();
-        auto iterator = info.find("version");
-        if (iterator != info.end()) {
-            versions.push_back(plugin->getName() + "_" + iterator->second);
+
+        // Does this plugin need secured data? If the plugin already has all of
+        // it's secure data, this should return 0. Otherwise, it should return
+        // the number of entries this plugin requires. A plugin will already have
+        // it's secure data if the BedrockServer gets destroyed and re-created
+        // but the binary is not terminated. This happens when it gets interrupted to
+        // handle a signal (like a SIGHUP that triggers a DB backup). If they need
+        // secure data, we'll skip initializing them here then instead initialize
+        // them after we've loaded the data.
+        list<string> needsSecureData = plugin->needsSecureData();
+        if (needsSecureData.size()) {
+            pluginSecureDataMap.insert(make_pair(plugin, needsSecureData));
+        } else {
+            plugin->initialize(args, *this);
+            plugins.push_back(plugin);
+
+            // If the plugin has version info, add it to the list.
+            auto info = plugin->getInfo();
+            auto iterator = info.find("version");
+            if (iterator != info.end()) {
+                versions.push_back(plugin->getName() + "_" + iterator->second);
+            }
+        }
+    }
+
+    // If any plguins needs secure data, load that now, then initialize.
+    if (!pluginSecureDataMap.empty()) {
+        _loadSecureData(pluginSecureDataMap, args);
+
+        for (auto plugin : pluginSecureDataMap) {
```

Comment:
Why not just `auto plugin : plugins` and do them all here? Then we can remove the repeated code in the `else` block above? (the `if` three lines up would only wrap the `_loadSecureData` call)

Reply:
Well you have to call `_loadSecureData` before you call initialize, so we can't initialize all of the plugins above and then call _loadSecureData.

Reply:
Yeah, the idea is don't do it above, just do it here.